### ISSUE TITLE
OpenPDF Not Using JVM Supplied Fonts

### ISSUE THREAD

**Issue by Sanadis** [LibrePDF/OpenPDF#494]
**Describe the bug**
We recently switched our JDK to OpenJDK and were looking to substitute the fonts with Liberation as it is a quick package install on Linux. The reporting tool we use is JasperReports which calls in to OpenPDF to find its fonts but it wasn't finding them no matter what. After a bit of digging through the sources, it seems like the list of available fonts has been hard coded using a completely unrelated framework to what the JVM is doing. I may be misunderstanding this greatly due to my limited knowledge in this area, but is there a reason OpenPDF doesn't use java.awt for the font lookup? The JVM already does all the leg work for loading in fonts via the OS and the corresponding fontconfig if they exist.

I know this issue should be fairly easily addressed by providing a JAR for JasperReports to handle the font extensions, however we do have a large number of releases and servers that make this more of an issue.

**To Reproduce**
This is a basic test to see what is available to our JVM runtime (These are groovy scripts I am testing with):
```
import java.awt.*;
final Font[] fonts = GraphicsEnvironment.getLocalGraphicsEnvironment().getAllFonts();
System.out.println("");
for (int i = 0; i < fonts.length; i++) {
  System.out.println(fonts[i].getFontName() + " : " + fonts[i].getFamily() + " : " + fonts[i].getName());
}
```

And some logic to test the matcher:
```
import java.awt.*;

matchFont("Liberation Sans", Font.PLAIN);
matchFont("Liberation Serif", Font.PLAIN);
matchFont("Liberation Mono", Font.PLAIN);
matchFont("DejaVu Sans", Font.PLAIN);
matchFont("DejaVu Serif", Font.PLAIN);
matchFont("DejaVu Sans Mono", Font.PLAIN);
matchFont("Helvetica", Font.PLAIN);

void matchFont(String fontName, int fontType) {
  Font font = new Font(fontName, fontType, 10);
  System.out.println("testFontName=" + fontName + ": matchFamily=" + font.getFamily()+", matchName=" + font.getFontName());
}
```

Output:
```
testFontName=Liberation Sans: matchFamily=Liberation Sans, matchName=Liberation Sans
testFontName=Liberation Serif: matchFamily=Liberation Serif, matchName=Liberation Serif
testFontName=Liberation Mono: matchFamily=Liberation Mono, matchName=Liberation Mono
testFontName=DejaVu Sans: matchFamily=DejaVu Sans, matchName=DejaVu Sans
testFontName=DejaVu Serif: matchFamily=Dialog, matchName=Dialog.plain
testFontName=DejaVu Sans Mono: matchFamily=Dialog, matchName=Dialog.plain
testFontName=Helvetica: matchFamily=SansSerif, matchName=SansSerif.plain
```

A couple of the fonts aren't available or configured so it just matches to a generic one.

Here is my attempt as OpenPDF matching:
```
import com.lowagie.text.*;

matchFont("Liberation Sans");
matchFont("Liberation Serif");
matchFont("Liberation Mono");
matchFont("DejaVu Sans");
matchFont("DejaVu Serif");
matchFont("DejaVu Sans Mono");
matchFont("Helvetica");
matchFont("Courier");

void matchFont(String fontName) {
  Font font = FontFactory.getFont(fontName, 10);
  System.out.println("testFontName=" + fontName + ", matchFamilyName=" + font.getFamilyname() + ", baseFont=" + (font.getBaseFont() ? font.getFamilyname() : null));
}
```

Output:
```
testFontName=Liberation Sans, matchFamilyName=unknown, baseFont=null
testFontName=Liberation Serif, matchFamilyName=unknown, baseFont=null
testFontName=Liberation Mono, matchFamilyName=unknown, baseFont=null
testFontName=DejaVu Sans, matchFamilyName=unknown, baseFont=null
testFontName=DejaVu Serif, matchFamilyName=unknown, baseFont=null
testFontName=DejaVu Sans Mono, matchFamilyName=unknown, baseFont=null
testFontName=Helvetica, matchFamilyName=Helvetica, baseFont=Helvetica
testFontName=Courier, matchFamilyName=Courier, baseFont=Courier
```

Based on the FontFactoryImp, the list of fonts appears to be hard coded if you are not manually registering them as a developer.

This is likely not a bug and more of a design issue which would be more of a feature request, but I want to verify if I am missing anything as a part of this process of adding more fonts.

Any assistance or input on this would be greatly appreciated.

**Comment by asturio** [comment]
I was searching for fonts in the JDK (8, and 11) and there are no Fonts in these. There is some mechanism for finding and using fonts of the OS (Linux uses e.g. fontconfig). Maybe we can use that mechanism in OpenPDF. Maybe in a new utility class "OsFonts" or the like.

The point is, the TTF-File is needed to instantiate an OS-Font in OpenPDF. And I haven't found a way to locate were these files are (from Java).

There is an `openpdf-fonts-extra` maven module, which contains Liberation fonts, and you could use it.

The solution using OS-Fonts would be nice, though.

**Comment by asturio** [comment]
Pull Requests are welcome!

**Comment by Sanadis** [comment]
That is correct, the JDK uses fontconfig based on the following documentation:
https://docs.oracle.com/javase/8/docs/technotes/guides/intl/fontconfig.html

However even without a $JDKHOME/lib/fontconfig, there still seems to be something in the JDK framework that when running on a Linux platform, it will load in the OS fontconfig if it exists. But regardless of that, the JDK seems to already have logic to load in a fontconfig and do matching without having to implement a separate system. Having never worked with these systems in the past I do not know what is needed to generate a font, if you need the TTF file or if the JDK has a way of generating it without needing the file locations.

Our current workaround for us is putting together a font extension library that the JasperReports framework loads in separately from OpenPDF.

One thing I am not certain on is if the VM options java.awt.headless=true has any impact on how fonts are handled by the JDK. 

I will do some digging and find out if it is possible to tie these components together.
